libcrux-ml-dsa/src/matrix.rs

```diff
@@ -114,27 +126,37 @@ pub(crate) fn compute_w_approx<
     const COLUMNS_IN_A: usize,
 >(
     A_as_ntt: &[[PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A]; ROWS_IN_A],
-    signer_response: [PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A],
+    mut signer_response: [PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A],
     verifier_challenge_as_ntt: PolynomialRingElement<SIMDUnit>,
     t1: [PolynomialRingElement<SIMDUnit>; ROWS_IN_A],
 ) -> [PolynomialRingElement<SIMDUnit>; ROWS_IN_A] {
     let mut result = [PolynomialRingElement::<SIMDUnit>::ZERO(); ROWS_IN_A];
 
-    for (i, row) in A_as_ntt.iter().enumerate() {
-        for (j, ring_element) in row.iter().enumerate() {
-            let product = ntt_multiply_montgomery(&ring_element, &ntt(signer_response[j]));
+    // Move signer response into NTT
+    for i in 0..signer_response.len() {
+        signer_response[i] = ntt(signer_response[i]);
+    }
 
-            result[i] = PolynomialRingElement::<SIMDUnit>::add(&result[i], &product);
+    cloop! {
+        for (i, row) in A_as_ntt.iter().enumerate() {
+            cloop! {
+                for (j, ring_element) in row.iter().enumerate() {
+                    let product = ntt_multiply_montgomery(&ring_element, &signer_response[j]);
+
+                    result[i] = PolynomialRingElement::<SIMDUnit>::add(&result[i], &product);
+                }
+            }
+
+            let t1_shifted =
+                shift_left_then_reduce::<SIMDUnit, { BITS_IN_LOWER_PART_OF_T as i32 }>(t1[i]);
+            let t1_shifted = ntt(t1_shifted);
+            let challenge_times_t1_shifted =
+                ntt_multiply_montgomery(&verifier_challenge_as_ntt, &t1_shifted);
+            result[i] = invert_ntt_montgomery(PolynomialRingElement::<SIMDUnit>::subtract(
+                &result[i],
+                &challenge_times_t1_shifted,
```

Comment:
I wonder if it would make sense, from a performance perspective, to move this to a separate loop before this one, to have a bunch of ntt loops in sequence. I have a feeling that might make sense from a caching perspective.

Reply:
this is just a comment, not a request to change the code

Reply:
Yes, it may. It's something to play with. Sometimes it makes a difference, sometimes it doesn't (when the compiler was intelligent enough already).